packages/@aws-cdk/pipelines/lib/docker-registry.ts

```diff
@@ -0,0 +1,164 @@
+}
+
+/** Docstrings */
+export interface EcrDockerRegistryOptions {
+  /** Docstrings */
+  readonly assumeRole?: iam.IRole
+  /** Docstrings */
+  readonly usages?: DockerRegistryUsage[];
+}
+
+/** Docstrings */
+export enum DockerRegistryUsage {
+  SYNTH,
+  SELF_UPDATE,
+  ASSET_PUBLISHING,
+};
+
+/** Docstrings */
+class ExternalDockerRegistry extends DockerRegistry {
+  constructor(public readonly registryDomain: string, private readonly opts: ExternalDockerRegistryOptions) {
+    super();
+  }
+
+  public grantRead(grantee: iam.IGrantable) {
+    if (this.opts.assumeRole) {
+      grantee.grantPrincipal.addToPrincipalPolicy(new iam.PolicyStatement({
+        actions: ['sts:AssumeRole'],
+        resources: [this.opts.assumeRole.roleArn],
+      }));
+    }
+    const role = this.opts.assumeRole ?? grantee;
+    this.opts.secret.grantRead(role);
+  }
+
+  /**
+   * Docstrings
+   * @internal
+   */
+  public _renderCdkAssetsConfig(): DockerRegistryCredentialSource {
+    return {
+      secretsManagerSecretId: this.opts.secret.secretArn,
+      secretsUsernameField: this.opts.secretUsernameField,
+      secretsPasswordField: this.opts.secretPasswordField,
+      assumeRoleArn: this.opts.assumeRole?.roleArn,
+    };
+  }
+}
+
+/** Docstrings */
+class EcrDockerRegistry extends DockerRegistry {
+  public readonly registryDomain: string;
+
+  constructor(private readonly repositories: ecr.IRepository[], private readonly opts: EcrDockerRegistryOptions) {
+    super();
+
+    if (repositories.length === 0) {
+      throw new Error('must supply at least one `ecr.IRepository` to create an `EcrDockerRegistry`');
+    }
+    this.registryDomain = Fn.select(0, Fn.split('/', repositories[0].repositoryUri));
+  }
+
+  public grantRead(grantee: iam.IGrantable) {
+    if (this.opts.assumeRole) {
+      grantee.grantPrincipal.addToPrincipalPolicy(new iam.PolicyStatement({
+        actions: ['sts:AssumeRole'],
+        resources: [this.opts.assumeRole.roleArn],
+      }));
+    }
+    const role = this.opts.assumeRole ?? grantee;
+    this.repositories.forEach(repo => repo.grantPull(role));
+  }
+
+  /**
+   * does stuff
+   * @internal
+   */
+  public _renderCdkAssetsConfig(): DockerRegistryCredentialSource {
+    return {
+      ecrRepository: true,
+      assumeRoleArn: this.opts.assumeRole?.roleArn,
+    };
+  }
+}
+
+interface DockerRegistryCredentialSource {
+  readonly secretsManagerSecretId?: string;
+  readonly secretsUsernameField?: string;
+  readonly secretsPasswordField?: string;
+  readonly ecrRepository?: boolean;
+  readonly assumeRoleArn?: string;
+}
+
+export function dockerRegistriesInstallCommands(registries?: DockerRegistry[], osType?: ec2.OperatingSystemType): string[] {
+  if (!registries || registries.length === 0) { return []; }
+
+  const domainCredentials = registries.reduce(function(map: Record<string, any>, registry) {
+    map[registry.registryDomain] = registry._renderCdkAssetsConfig();
```

Comment:
We could do without making `registryDomain` public if we did:

```ts
public _renderCdkAssetConfig() {
  return { [this.registryDomain]: { ...whatevs ... }};
}

// and here
Object.assign(map, registry._renderCdkAssetsConfig());
```

Just a thought

Reply:
I like it. I was trying to do something similar, but couldn't get the syntax just right.